client/gui_rpc_server_ops.cpp

```diff
@@ -1289,6 +1291,79 @@ static bool complete_post_request(char* buf) {
     return true;
 }
 
+// We use a different authentication scheme for HTTP because
+// each request has its own connection.
+// Send clients an "authentication ID".
+// Each request has (in HTTP header vars) the ID,
+// a sequence number, and a hash of the seq# and the password.
+//
+struct AUTH_INFO {
+    int id;
+    long seqno;
+    char salt[64];
+};
+
+vector<AUTH_INFO> auth_infos;
+
+// check HTTP authentication info
+//
+bool valid_auth(int id, long seqno, char* hash, char* request) {
+    char buf[256], my_hash[256];
+    //printf("valid_auth: id %d seqno %ld hash %s\n", id, seqno, hash);
+    for (unsigned int i=0; i<auth_infos.size(); i++) {
+        AUTH_INFO& ai = auth_infos[i];
+        if (ai.id != id) continue;
+        if (seqno <= ai.seqno) return false;
+        int n = request?(int)strlen(request):0;
+        snprintf(buf, sizeof(buf), "%ld%s%s", seqno, gstate.gui_rpcs.password, ai.salt);
+        md5_block((const unsigned char*)buf, (int)strlen(buf), my_hash,
+            (const unsigned char*)request, n
+        );
+        if (strcmp(hash, my_hash)) {
+            msg_printf(0, MSG_INFO, "got invalid GUI RPC request");
+            return false;
+        }
+        ai.seqno = seqno;   // bump seqno only if valid request
+        return true;
+    }
+    return false;
+}
+
+// create a new authentication ID
+//
+void handle_get_auth_id(MIOFILE& fout) {
+    static int id=0;
+    AUTH_INFO ai;
+    ai.id = id++;
+    ai.seqno = 0;
+    make_random_string(ai.salt);
+    auth_infos.push_back(ai);
+    fout.printf("<auth_id>%d</auth_id>\n<auth_salt>%s</auth_salt>\n", ai.id, ai.salt);
+}
+
+// see if the HTTP request has valid authentication info
+//
+static bool authenticated_request(char* buf) {
+    int auth_id;
+    long auth_seqno;
+    char auth_hash[256];
+    char* p = strstr(buf, "Auth-ID: ");
+    if (!p) return false;
+    int n = sscanf(p+strlen("Auth-ID: "), "%d", &auth_id);
+    if (n != 1) return false;
+    p = strstr(buf, "Auth-Seqno: ");
+    if (!p) return false;
+    n = sscanf(p+strlen("Auth-Seqno: "), "%ld", &auth_seqno);
+    if (n != 1) return false;
+    p = strstr(buf, "Auth-Hash: ");
+    if (!p) return false;
+    n = sscanf(p+strlen("Auth-Hash: "), "%64s", auth_hash);
+    if (n != 1) return false;
+    char* request = strstr(buf, "\r\n\r\n");
+    if (request) request += 4;
```

Comment:
Can `"\r\n\r\n"` be moved to constant and it size be added to `request` variable instead of using magic numbers?
If for some reasons `"\r\n\r\n"` will be changed to. for example, `"\r\n"` it's very easy to forget to change magic number `4` to `2`